packages/context/src/app/AppProvider.tsx

```diff
@@ -62,8 +84,48 @@ export type AppProviderProps = {
 
 function AppProvider(props: AppProviderProps) {
     const { app, children } = props;
-
-    return <AppContext.Provider value={app}>{children}</AppContext.Provider>;
+    const apiServices = useApiServices();
+
+    const appWithFunctions: AppExtended = {
+        ...app,
+        findDefaultPageSlugFor: useCallback(
+            async (objectId: string) => {
+                let defaultPageId: string | undefined;
+                let currentObjectId: string | undefined = objectId;
+                while (currentObjectId !== undefined) {
+                    if (app.defaultPages?.[currentObjectId]) {
+                        defaultPageId = app.defaultPages[currentObjectId];
+                        break;
+                    }
+
+                    const effectiveObject: Obj | undefined = await apiServices.get<Obj>(
+                        `data/objects/${currentObjectId}/effective`,
+                        {
+                            params: { filter: { fields: ['baseObject'] } },
+                        },
+                    );
+
+                    currentObjectId = effectiveObject?.baseObject?.objectId ?? undefined;
+                }
+
+                let defaultPage: Page | undefined;
+                if (defaultPageId) {
+                    const pageId = defaultPageId.includes('/')
+                        ? defaultPageId.split('/').slice(2).join('/')
+                        : defaultPageId;
+                    defaultPage = await apiServices.get<Page>(
```

Comment:
I think we can catch 404 errors here.  If the page doesn't exist, probably shouldn't make the widget fail, just treat it as if there is no default page.